### DISCUSSION TITLE
Logging Notification Adapter

### DISCUSSION THREAD

**Discussion by MannTec** [qutebrowser/qutebrowser#6398] [Q&A] [answered]
Hello Everyone,

I would like to know if anyone can provide an example of how to add a notification adapter that saves directly to a text file? Something similar to the current **MessagesNotificationAdapter** function?

**Comment by The-Compiler** [comment, marked as answer]
What's your use-case? Maybe something like [tiramisu](https://github.com/Sweets/tiramisu) would already do what you want, with the benefit that it'd work with other notifications (i.e. system-wide) as well?

**Reply by MannTec** [reply]
The use-case would be allowing the user to log each notification to a text file to build a history of the frequency of notifications from popular sites around the web. User would login a popular website that supports notifications, once a notification is received the data is appended to a text file, the user then collects the data later for analysis.

The MessagesNotificationAdapter is almost exactly what I need but instead for sending a message to be displayed, it would append the information to a local file. No graphical representation or OS toast wanted/needed.

I've used Tiramisu in the past and it seems to have issues with Chromium on Windows 10 that I'm sure I can avoid using QB. If it wasn't for those issues it would be a perfect fit. I appreciate any advice you can provide. Thank you.

**Reply by The-Compiler** [reply]
That's due to https://github.com/Sweets/tiramisu/issues/20 - qutebrowser [works around it](https://github.com/qutebrowser/qutebrowser/blob/v2.2.0/qutebrowser/browser/webengine/notification.py#L794-L796), so it should work on Linux. No idea about Windows 10 though (with WSL).

As for your original question, not sure what more I can say. Write a subclass of `AbstractNotificationAdapter` which implements the abstract methods, similar to what `MessagesNotificationAdapter` does. Not sure I'd accept something like that in qutebrowser itself, though.

**Comment by MannTec** [comment]
These changes are isolated and academic. My Python knowledge is far below the complexity of the Qutebrowser code and I wanted expert opinion on this modification. If the scope is confined to just adding a new subclass I will mark this as answered and you have my gratitude.

**Reply by The-Compiler** [reply]
That's almost all - add a new subclass, add a new value in `configdata.yml` and adjust `NotificationBridgePresenter._init_adapter` to use it. See e.g. 356a170ce30e94e42534ecbded402af852c5a67d or e2e4c74ffb0df411cca76be55cdebff65de44c19 (simpler but slightly outdated) for some inspiration.
